File: kafka-topic-summary.js
const fs = require('fs');
const Kafka = require('node-rdkafka');

const configFile = 'config.json';

const loadConfig = () => {
    const configData = fs.readFileSync(configFile);
    return JSON.parse(configData);
};

const createConsumer = (config) => {
    const consumer = new Kafka.KafkaConsumer(config.consumer);

    consumer.setKafkaConfig(config.kafka);

    consumer.on('event.error', (err) => {
        console.error('Error:', err);
        consumer.disconnect();
    });

    return consumer;
};

const createProducer = (config) => {
    const producer = new Kafka.Producer(config.producer);

    producer.setKafkaConfig(config.kafka);

    producer.on('event.error', (err) => {
        console.error('Error:', err);
        producer.disconnect();
    });

    return producer;
};

const getTopicMetadata = (config, callback) => {
    const consumer = createConsumer(config);

    consumer.connect();

    consumer.on('ready', () => {
        const metadata = consumer.getMetadata();
        consumer.disconnect();
        callback(metadata);
    });
};

const getTimestampsForTopic = (config, topic) => {
    return new Promise((resolve, reject) => {
        getTopicMetadata(config, (metadata) => {
            const topicMetadata = metadata.topics.find((t) => t.name === topic);

            if (!topicMetadata) {
                reject(new Error(`Topic '${topic}' not found in metadata`));
                return;
            }

            const timestamps = [];
            const partitions = topicMetadata.partitions;

            const consumer = createConsumer(config);

            consumer.connect();

            consumer.on('ready', () => {
                const fetchTimestamps = () => {
                    let completedPartitions = 0;

                    partitions.forEach((partition) => {
                        consumer.offsetsForTime(topic, partition.id, -1, 1, (err, offsets) => {
                            if (err) {
                                reject(err);
                                return;
                            }

                            if (offsets && offsets.length > 0) {
                                const offset = offsets[0];
                                if (offset.offset !== Kafka.OFFSET_INVALID) {
                                    const timestamp = new Date(offset.timestamp * 1000);
                                    timestamps.push(timestamp);
                                }
                            }

                            completedPartitions++;
                            if (completedPartitions === partitions.length) {
                                consumer.disconnect();
                                resolve(timestamps);
                            }
                        });
                    });
                };

                fetchTimestamps();
            });

            consumer.on('event.error', (err) => {
                reject(err);
                consumer.disconnect();
            });
        });
    });
};

const countMessagesForDay = (config, topic, consumerGroup, date) => {
    return new Promise((resolve, reject) => {
        const consumerConfig = Object.assign({}, config.consumer, { 'group.id': consumerGroup });
        const consumer = createConsumer({ consumer: consumerConfig, kafka: config.kafka });

        consumer.connect();

        consumer.on('ready', () => {
            consumer.subscribe([topic]);
            consumer.consume();
        });

        let totalCount = 0;
        let totalSize = 0;

        consumer.on('data', (message) => {
            const messageTime = new Date(message.timestamp);
            if (messageTime >= date && messageTime < new Date(date.getTime() + 24 * 60 * 60 * 1000)) {
                totalCount++;
                totalSize += message.value ? message.value.length : 0;
            }
        });

        // ...

        consumer.on('event.error', (err) => {
            reject(err);
            consumer.disconnect();
        });

        consumer.on('disconnected', () => {
            const averageSize = totalCount > 0 ? totalSize / totalCount : 0;
            resolve({ count: totalCount, size: totalSize, averageSize });
        });
    });
};

const generateReport = async (config) => {
    const topics = config.topics;
    const consumerGroups = config.consumerGroups;

    for (const topic of topics) {
        for (const consumerGroup of consumerGroups) {
            const timestamps = await getTimestampsForTopic(config, topic);
            console.log(`Topic: ${topic}, Consumer Group: ${consumerGroup}`);
            console.log('--------------------------');
            for (const timestamp of timestamps) {
                const result = await countMessagesForDay(config, topic, consumerGroup, timestamp);
                console.log(`Date: ${timestamp.toISOString().slice(0, 10)}`);
                console.log(`Total Count: ${result.count}`);
                console.log(`Total Size: ${result.size}`);
                console.log(`Average Message Size: ${result.averageSize}`);
                console.log('--------------------------');
            }
        }
    }
};

const config = loadConfig();
generateReport(config);

File: kafka-topic-summary.py
import json
from datetime import datetime, timedelta
from confluent_kafka import Consumer, KafkaException

def load_config():
    with open('config.json') as config_file:
        config_data = json.load(config_file)
    return config_data

def count_messages_for_day(config, topic, consumer_group, date):
    consumer_config = {
        'bootstrap.servers': config['kafka']['bootstrap.servers'],
        'group.id': consumer_group,
        'auto.offset.reset': 'earliest',
        'enable.auto.commit': False,
    }
    consumer = Consumer(consumer_config)
    
    consumer.subscribe([topic])
    
    total_count = 0
    total_size = 0
    
    try:
        while True:
            message = consumer.poll(timeout=1.0)
            if message is None:
                continue
            if message.error():
                raise KafkaException(message.error())
            
            message_time = datetime.utcfromtimestamp(message.timestamp() / 1000)
            if message_time.date() == date:
                total_count += 1
                total_size += len(message.value() or b'')
            
            consumer.commit(asynchronous=False)
    except KeyboardInterrupt:
        pass
    finally:
        consumer.close()
    
    return {'count': total_count, 'size': total_size, 'average_size': total_size / total_count if total_count > 0 else 0}

def generate_report(config):
    topics = config['topics']
    consumer_groups = config['consumerGroups']

    for topic in topics:
        for consumer_group in consumer_groups:
            timestamps = get_timestamps_for_topic(config, topic)
            print(f'Topic: {topic}, Consumer Group: {consumer_group}')
            print('--------------------------')
            for timestamp in timestamps:
                result = count_messages_for_day(config, topic, consumer_group, timestamp.date())
                print(f"Date: {timestamp.date().isoformat()}")
                print(f"Total Count: {result['count']}")
                print(f"Total Size: {result['size']}")
                print(f"Average Message Size: {result['average_size']}")
                print('--------------------------')

def get_timestamps_for_topic(config, topic):
    consumer_config = {
        'bootstrap.servers': config['kafka']['bootstrap.servers'],
        'group.id': 'timestamp_group',
        'auto.offset.reset': 'earliest',
        'enable.auto.commit': False,
    }
    consumer = Consumer(consumer_config)
    
    consumer.subscribe([topic])
    partitions = consumer.assignment()
    
    timestamps = []
    
    try:
        for partition in partitions:
            low, high = consumer.get_watermark_offsets(partition)
            start_offset = consumer.offsets_for_times([(partition, low)], timeout=1.0)[0][1].offset
            end_offset = consumer.offsets_for_times([(partition, high)], timeout=1.0)[0][1].offset
            
            if start_offset is None or end_offset is None:
                continue
            
            consumer.seek(partition, start_offset)
            while consumer.position(partition) < end_offset:
                msg = consumer.poll(timeout=1.0)
                if msg is None:
                    continue
                if msg.error():
                    raise KafkaException(msg.error())
                timestamps.append(datetime.utcfromtimestamp(msg.timestamp() / 1000))
                consumer.commit(asynchronous=False)
            consumer.unsubscribe()
    except KeyboardInterrupt:
        pass
    finally:
        consumer.close()
    
    return timestamps

config = load_config()
generate_report(config)
